code/go/0chain.net/smartcontract/zcnsc/models.go

```diff
@@ -0,0 +1,497 @@
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// To review: tokenLock init values
+func getNewAuthorizer(pk string, id string) *authorizerNode {
+	return &authorizerNode{
+		PublicKey: pk,
+		Staking: &tokenpool.ZcnLockingPool{
+			ZcnPool: tokenpool.ZcnPool{
+				TokenPool: tokenpool.TokenPool{
+					ID:      id,
+					Balance: 0,
+				},
+			},
+			TokenLockInterface: tokenLock{
+				StartTime: 0,
+				Duration:  0,
+				Owner:     id,
+			},
+		},
+		ID: id,
+	}
+}
+
+type authorizerNodes struct {
+	NodeMap map[string]*authorizerNode `json:"node_map"`
+}
+
+func (an *authorizerNodes) Decode(input []byte) error {
+	if an.NodeMap == nil {
+		an.NodeMap = make(map[string]*authorizerNode)
+	}
+
+	var objMap map[string]json.RawMessage
+	err := json.Unmarshal(input, &objMap)
+	if err != nil {
+		return err
+	}
+
+	nodeMap, ok := objMap["node_map"]
+	if ok {
+		var authorizerNodes map[string]json.RawMessage
+		err := json.Unmarshal(nodeMap, &authorizerNodes)
+		if err != nil {
+			return err
+		}
+
+		for _, raw := range authorizerNodes {
+			target := &authorizerNode{}
+			err := target.Decode(raw, &tokenLock{})
+			if err != nil {
+				return err
+			}
+
+			err = an.addAuthorizer(target)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (an *authorizerNodes) Encode() []byte {
+	buff, _ := json.Marshal(an)
+	return buff
+}
+
+func (an *authorizerNodes) GetHash() string {
+	return util.ToHex(an.GetHashBytes())
+}
+
+func (an *authorizerNodes) GetHashBytes() []byte {
+	return encryption.RawHash(an.Encode())
+}
+
+func (an *authorizerNodes) save(balances cstate.StateContextI) (err error) {
+	_, err = balances.InsertTrieNode(allAuthorizerKey, an)
+	return
+}
+
+func (an *authorizerNodes) deleteAuthorizer(id string) (err error) {
+	if an.NodeMap[id] == nil {
+		err = common.NewError("failed to delete authorizer", fmt.Sprintf("authorizer (%v) does not exist", id))
+		return
+	}
+	delete(an.NodeMap, id)
+	return
+}
+
+func (an *authorizerNodes) addAuthorizer(node *authorizerNode) (err error) {
+	if node == nil {
+		err = common.NewError("failed to add authorizer", "authorizerNode is not initialized")
+		return
+	}
+
+	if an.NodeMap == nil {
+		err = common.NewError("failed to add authorizer", "receiver NodeMap is not initialized")
+		return
+	}
+
+	if an.NodeMap[node.ID] != nil {
+		err = common.NewError("failed to add authorizer", fmt.Sprintf("authorizer (%v) already exists", node.ID))
+		return
+	}
+
+	an.NodeMap[node.ID] = node
+
+	return
+}
+
+func (an *authorizerNodes) updateAuthorizer(node *authorizerNode) (err error) {
+	if an.NodeMap[node.ID] == nil {
+		err = common.NewError("failed to update authorizer", fmt.Sprintf("authorizer (%v) does not exist", node.ID))
+		return
+	}
+	an.NodeMap[node.ID] = node
+	return
+}
+
+func getAuthorizerNodes(balances cstate.StateContextI) (*authorizerNodes, error) {
+	an := &authorizerNodes{}
+	av, err := balances.GetTrieNode(allAuthorizerKey)
+	if err != nil {
```

Comment:
I think best style would be somthing like
```go
if err != nil {
  if err != unit.ErrValueNotPresent {
        return nil, err
  } else {
        an.NodeMap = make(map[string]*authorizerNode)
        return an, nil  
  }
}

```

Reply:
Fixed